pycloudlib/azure/cloud.py

```diff
@@ -1009,6 +1041,31 @@ def snapshot(
 
         return image_id
 
+    def get_boot_diagnostics(
```

Comment:
There's a lot going on here. You've exposed boot diagnostic retrieval as a public interface, but realistically the responsibilities should be split:

1. A private method which automatically tries to get the boot diagnostics for an instance as part of creation, saving the boot diagnostics in memory
2. A public interface for users to retrieve the diagnostics ad hoc

With the current implementation the class will constantly need to re-request the diagnostics along with all of the state management which comes with it. Additionally, this implementation requires the user to specify the instance they want to have boot diagnostics for, and arbitrarily returns the boot diagnostics for the first VM in the VM list if none is specified.

We care about boot diagnostics mainly for when things go wrong, which will likely happen when calling `launch()`. Ideally this method would be bound to a cloud's instance, since it is for a given instance we are retrieving the boot diagnostics for. Unfortunately the Azure SDK requires both the VM name and resource group to resolve the diagnostics, so you've correctly injected the instance as part of the method signature. However the user will only have access to the instance if `launch()` succeeds, hence your default for grabbing the first instance if none is specified. However, if users want to retrieve the logs for failed deployments after the first created instance they're out of luck, so the interface is now inconsistent.

A bit overly detailed but I wanted to be explicit in how the current interface is hiding too many details from the user where both the usage and results are cumbersome. Revisit this interface and work backwards from what we you want to expose to the user (return the logs) and determine a more appropriate means of mapping boot diagnostics to created instances - where the set of possibilities include:

- No running instances
- Boot diagnostics is disabled
- One running instance
- Many running instances
- Requesting for an instance that has failed to launch